rclcpp/src/rclcpp/typesupport_helpers.cpp

```diff
@@ -0,0 +1,145 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rclcpp/typesupport_helpers.hpp"
+
+#include <memory>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+#include <tuple>
+#include <utility>
+
+#include "ament_index_cpp/get_package_prefix.hpp"
+#include "ament_index_cpp/get_resources.hpp"
+#include "rcpputils/shared_library.hpp"
+#include "rcpputils/find_library.hpp"
+#include "rosidl_runtime_cpp/message_type_support_decl.hpp"
+
+namespace rclcpp
+{
+
+std::string get_typesupport_library_path(
+  const std::string & package_name, const std::string & typesupport_identifier)
+{
+  const char * dynamic_library_folder;
+#ifdef _WIN32
+  dynamic_library_folder = "/bin/";
+#elif __APPLE__
+  dynamic_library_folder = "/lib/";
+#else
+  dynamic_library_folder = "/lib/";
+#endif
+
+  std::string package_prefix;
+  try {
+    package_prefix = ament_index_cpp::get_package_prefix(package_name);
+  } catch (ament_index_cpp::PackageNotFoundError & e) {
+    throw std::runtime_error(e.what());
+  }
+
+  const std::string library_path = rclcpp::path_for_library(
+    package_prefix + dynamic_library_folder,
+    package_name + "__" + typesupport_identifier);
+  if (library_path.empty()) {
+    throw std::runtime_error(
+            "Typesupport library for " + package_name + " does not exist in '" + package_prefix +
+            "'.");
+  }
+  return library_path;
+}
+
+const std::pair<std::string, std::string> extract_type_and_package(const std::string & full_type)
+{
+  std::string package_name;
+  std::string type_name;
+
+  std::tie(package_name, std::ignore, type_name) = extract_type_identifier(full_type);
+
+  return {package_name, type_name};
+}
+
+std::tuple<std::string, std::string, std::string>
+extract_type_identifier(const std::string & full_type)
+{
+  char type_separator = '/';
+  auto sep_position_back = full_type.find_last_of(type_separator);
+  auto sep_position_front = full_type.find_first_of(type_separator);
+  if (sep_position_back == std::string::npos ||
+    sep_position_back == 0 ||
+    sep_position_back == full_type.length() - 1)
+  {
+    throw std::runtime_error(
+            "Message type is not of the form package/type and cannot be processed");
+  }
+
+  std::string package_name = full_type.substr(0, sep_position_front);
+  std::string middle_module = "";
+  if (sep_position_back - sep_position_front > 0) {
+    middle_module =
+      full_type.substr(sep_position_front + 1, sep_position_back - sep_position_front - 1);
+  }
+  std::string type_name = full_type.substr(sep_position_back + 1);
+
+  return std::make_tuple(package_name, middle_module, type_name);
+}
```

Comment:
These seem like ideal candidates for C++17's new `std::string_view`, e.g.:

```suggestion
std::pair<std::string_view, std::string_view>
extract_type_and_package(const std::string & full_type)
{
  std::string_view package_name;
  std::string_view type_name;

  std::tie(package_name, std::ignore, type_name) = extract_type_identifier(full_type);

  return {package_name, type_name};
}

std::tuple<std::string_view, std::string_view, std::string_view>
extract_type_identifier(std::string_view full_type)
{
  char type_separator = '/';
  auto sep_position_back = full_type.find_last_of(type_separator);
  auto sep_position_front = full_type.find_first_of(type_separator);
  if (sep_position_back == std::string::npos ||
    sep_position_back == 0 ||
    sep_position_back == full_type.length() - 1)
  {
    throw std::runtime_error(
            "Message type is not of the form package/type and cannot be processed");
  }

  std::string_view package_name = full_type.substr(0, sep_position_front);
  std::string_view middle_module = "";
  if (sep_position_back - sep_position_front > 0) {
    middle_module =
      full_type.substr(sep_position_front + 1, sep_position_back - sep_position_front - 1);
  }
  std::string_view type_name = full_type.substr(sep_position_back + 1);

  return std::make_tuple(package_name, middle_module, type_name);
}
```

I think that would avoid all of the copied strings. :)

Reply:
Obviously just a suggestion, not a requirement. Though it would be nice to eliminate any sources of memory allocation we can.

Reply:
I didn't know rclcpp uses C++17. I'd keep it string, since string_view is alleged to be a bit dangerous and I don't know it very well yet. Is the goal of avoiding allocations here performance or certifiability?

Reply:
It can use it now, as of rolling/galactic.

The only thing dangerous about it is that the string_view should not outlive the string it is viewing. In this case I think that would be safe, but it was just a suggestion. You can resolve it if you want.

Reply:
I do agree this is a neat case where string_view would fit, and in Rust I'd make it a &str. But I'm more comfortable with string here.